w3af/plugins/audit/hpp.py

```diff
@@ -0,0 +1,263 @@
+
+                i = Info(u'Input concatenation detected', desc, ptest_resp.id,
+                         self.get_name())
+                i.set_url(ptest_url)
+                i.precedence = CONCAT
+                self.kb_append(self, 'hpp', i)
+                return
+
+            om.out.debug(u'Unable to determine precedence for HPP detection '
+                         u'with this request %s. Presumably not vulnerable to'
+                         u' HPP.' % ptest_url)
+            return
+
+    def client_side(self, freq, orig_response, precedence, value, param):
+
+        reflections = find_reflection(orig_response, precedence, [value])
+
+        for reflection in reflections:
+            #getting the list of exploitable params
+            forgeable = get_forgeable(reflection, precedence)
+
+            for f in forgeable:
+                init_val = reflection[1].get_querystring()[f][0]
+                if init_val.isdigit():
+                    forged_val = str(int(init_val)+1)
+                else:
+                    forged_val = FORGE_TOKEN
+                #building attack payload and exploiting
+                payloads = [unicode(value) + '&' + f + '=' + forged_val]
+
+                mutants = create_mutants(freq, payloads,
+                                         fuzzable_param_list=param,
+                                         mutant_tuple=(QSMutant,))
+                mu = mutants[0]
+
+                vtest_resp = self._uri_opener.send_mutant(mu)
+                #checkig for successful exploitation
+                r = find_reflection(vtest_resp, precedence, [forged_val])
+                if r:
+                    msg = u"The %s uri is vulnerable to HPP attack: value of " \
+                          u"parameter %s can be forged using URL: %s, " \
+                          u"precedence is %s"
+                    msg = msg % (freq.get_uri(), r, mu.get_uri(), precedence)
+                    v = Vuln.from_mutant(u'HPP', desc=msg, mutant=mu,
+                                         plugin_name='hpp', response_ids=[],
+                                         severity=severity.MEDIUM)
+                    v.precedence = precedence
+                    v.set_url(freq.get_uri())
+                    # debug_msg = debug_msg % (rid, freq.get_uri())
+                    self.kb_append_uniq(self, 'hpp', v,)
+
+    def get_long_desc(self):
+        """
+        :return: A DETAILED description of the plugin functions and features.
+        """
+        return """
+        This plugin detects client-side HTTP Parameter Pollution
+        vulnerabilities using the following technique. First it detects if any
+        of the input params is being reflected. If value of such parameter
+        exists in any of the links of received HTML code it will try to forge
+        it using HPP attack according to determined precedence or just
+        report if precedence is concatenation.
```

Comment:
This plugin detects client-side HTTP Parameter Pollution vulnerabilities

The following technique is used:
 * Detect if any of the input params is being reflected
 * For each reflected parameter check if they are used in HTML links
 * Send an HPP check to inject an extra parameter in the link